encoding/thrift/inbound_nowire.go

```diff
@@ -0,0 +1,207 @@
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	res, err := decodeNoWireRequest(ctx, call, treq, rw, t.NoWireHandler, wire.Call, t.Protocol, t.Enveloping)
+	if err != nil {
+		return err
+	}
+
+	if resType := res.Body.EnvelopeType(); resType != wire.Reply {
+		return errors.ResponseBodyEncodeError(
+			treq, errUnexpectedEnvelopeType(resType))
+	}
+
+	if res.IsApplicationError {
+		rw.SetApplicationError()
+
+		if applicationErrorMetaSetter, ok := rw.(transport.ApplicationErrorMetaSetter); ok {
+			applicationErrorMetaSetter.SetApplicationErrorMeta(&transport.ApplicationErrorMeta{
+				Details: res.ApplicationErrorDetails,
+				Name:    res.ApplicationErrorName,
+				Code:    res.ApplicationErrorCode,
+			})
+		}
+	}
+
+	if err := call.WriteToResponse(rw); err != nil {
+		// not reachable
+		return err
+	}
+
+	if err := res.ResponseWriter.WriteResponse(wire.Reply, rw, res.Body); err != nil {
+		return errors.ResponseBodyEncodeError(treq, err)
+	}
+
+	return nil
+}
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	_, err := decodeNoWireRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
+	return err
+}
+
+// decodeNoWireRequest is a shared utility between the implementations of
+// transport.UnaryHandler and transport.OnewayHandler, to decode and execute the
+// request regardless of enveloping, via the "nowire" implementation in
+// ThriftRW.
+func decodeNoWireRequest(
+	ctx context.Context,
+	call *encodingapi.InboundCall,
+	treq *transport.Request,
+	rw transport.ResponseWriter,
+	noWireHandler NoWireHandler,
+	reqEnvelopeType wire.EnvelopeType,
+	proto stream.Protocol,
+	enveloping bool,
+) (NoWireResponse, error) {
+	if err := errors.ExpectEncodings(treq, Encoding); err != nil {
+		return _emptyResponse, err
+	}
+
+	if err := call.ReadFromRequest(treq); err != nil {
+		return _emptyResponse, err
+	}
+
+	nwc := NoWireCall{
+		Reader:       treq.Body,
+		EnvelopeType: reqEnvelopeType,
+	}
+
+	if reqReader, ok := proto.(stream.RequestReader); ok {
+		nwc.RequestReader = reqReader
+	} else {
+		nwc.RequestReader = &reqReaderProto{
+			Protocol:   proto,
+			treq:       treq,
+			enveloping: enveloping,
+		}
+	}
+	return noWireHandler.Handle(ctx, &nwc)
+}
+
+// reqReaderProto is an implementation of ThriftRW's stream.RequestReader in
+// case the provided stream.Protocol does not implement the necessary
+// `ReadRequest` to discover the correct enveloping.
+type reqReaderProto struct {
+	stream.Protocol
+
+	treq       *transport.Request
+	enveloping bool
+}
+
+var _ stream.RequestReader = (*reqReaderProto)(nil)
+
+func (p *reqReaderProto) ReadRequest(
+	ctx context.Context,
+	et wire.EnvelopeType,
+	r io.Reader,
+	body stream.BodyReader,
+) (stream.ResponseWriter, error) {
+	sr := p.Protocol.Reader(r)
+	defer sr.Close()
+
+	if p.enveloping {
```

Comment:
I'm not a big fan of two code paths that are all but completely different are inside the same type with a bool switch. The switch should be in decodeNoWireResponse, constructing either an enveloped variant of this type, or a non-enveloped.

Fixing.